akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -253,7 +262,8 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
   private val taskRunner = new TaskRunner(system)
 
   // FIXME: This does locking on putIfAbsent, we need something smarter
-  private[this] val associations = new ConcurrentHashMap[Address, Association]()
+  private[this] val associationsByAddress = new ConcurrentHashMap[Address, Association]()
```

Comment:
I think it is time to move this to a class in itself, so it can make sure its invariants are kept intact (updating the two map properly). I also feel a risk of races here, accessing the two maps concurrently.